// express/index.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const http = require('node:http');
const { URL } = require('node:url');

/**
 * Creates a minimal Express-compatible app for the test suite.
 * @returns {Function & { use: Function, get: Function, post: Function, disable: Function, listen: Function }} App instance.
 */
function createExpress() {
  const middlewares = [];
  const routes = [];
  const disabledSettings = new Set();

  /**
   * Adds a middleware or route handler.
   * @param {string|Function} pathOrHandler - Middleware path or handler.
   * @param {...Function} handlers - Middleware handlers.
   * @returns {Function} App.
   */
  function use(pathOrHandler, ...handlers) {
    if (typeof pathOrHandler === 'string') {
      const basePath = pathOrHandler;
      for (const handler of handlers) {
        middlewares.push({ basePath, handler });
      }
      return app;
    }

    const allHandlers = [pathOrHandler, ...handlers];
    for (const handler of allHandlers) {
      middlewares.push({ basePath: '', handler });
    }
    return app;
  }

  /**
   * Registers a route handler.
   * @param {string} method - HTTP method.
   * @param {string} routePath - Route path.
   * @param {Function} handler - Route handler.
   * @returns {Function} App.
   */
  function registerRoute(method, routePath, handler) {
    routes.push({
      method,
      routePath,
      segments: routePath.split('/').filter(Boolean),
      handler,
    });
    return app;
  }

  /**
   * Disables a named setting.
   * @param {string} name - Setting name.
   * @returns {Function} App.
   */
  function disable(name) {
    disabledSettings.add(String(name));
    return app;
  }

  /**
   * Starts the app on an HTTP server.
   * @param {...any} args - Listen arguments.
   * @returns {http.Server} Server instance.
   */
  function listen(...args) {
    const server = http.createServer((req, res) => {
      handleRequest(req, res).catch((error) => {
        if (!res.headersSent) {
          res.statusCode = error?.status && Number.isInteger(error.status) ? error.status : 500;
          res.setHeader('Content-Type', 'application/json; charset=utf-8');
        }
        if (!res.writableEnded) {
          res.end(JSON.stringify({
            ok: false,
            error: {
              message: error instanceof Error ? error.message : String(error),
            },
          }));
        }
      });
    });

    return server.listen(...args);
  }

  /**
   * Handles a single request through middleware and route matching.
   * @param {http.IncomingMessage} req - Request.
   * @param {http.ServerResponse} res - Response.
   */
  async function handleRequest(req, res) {
    enhanceResponse(res);
    const parsedUrl = new URL(req.url || '/', 'http://127.0.0.1');
    req.path = parsedUrl.pathname;
    req.query = Object.fromEntries(parsedUrl.searchParams.entries());

    let middlewareIndex = 0;

    /**
     * Runs the next middleware or route.
     * @returns {Promise<void>}
     */
    async function next() {
      while (middlewareIndex < middlewares.length) {
        const entry = middlewares[middlewareIndex++];
        if (entry.basePath && !req.path.startsWith(entry.basePath)) {
          continue;
        }

        const maybePromise = entry.handler(req, res, next);
        if (maybePromise && typeof maybePromise.then === 'function') {
          await maybePromise;
        }
        return;
      }

      for (const route of routes) {
        if (route.method !== req.method) {
          continue;
        }

        const params = matchRoute(route.segments, req.path);
        if (!params) {
          continue;
        }

        req.params = params;
        const maybePromise = route.handler(req, res, next);
        if (maybePromise && typeof maybePromise.then === 'function') {
          await maybePromise;
        }
        return;
      }

      if (!res.writableEnded) {
        res.statusCode = 404;
        res.end('Not Found');
      }
    }

    await next();
  }

  /**
   * Enhances a response with Express-like helpers.
   * @param {http.ServerResponse} res - Response.
   */
  function enhanceResponse(res) {
    if (typeof res.status !== 'function') {
      res.status = function status(code) {
        res.statusCode = code;
        return res;
      };
    }

    if (typeof res.set !== 'function') {
      res.set = function set(field, value) {
        if (typeof field === 'string') {
          res.setHeader(field, value);
        } else if (field && typeof field === 'object') {
          for (const [key, headerValue] of Object.entries(field)) {
            res.setHeader(key, headerValue);
          }
        }
        return res;
      };
    }

    if (typeof res.json !== 'function') {
      res.json = function json(body) {
        if (!res.headersSent) {
          res.setHeader('Content-Type', 'application/json; charset=utf-8');
        }
        res.end(JSON.stringify(body));
        return res;
      };
    }

    if (typeof res.send !== 'function') {
      res.send = function send(body) {
        if (Buffer.isBuffer(body) || typeof body === 'string') {
          res.end(body);
          return res;
        }
        return res.json(body);
      };
    }

    if (typeof res.sendFile !== 'function') {
      res.sendFile = function sendFile(filePath) {
        const data = fs.readFileSync(filePath);
        if (!res.headersSent) {
          res.setHeader('Content-Type', contentTypeForFile(filePath));
        }
        res.end(data);
        return res;
      };
    }

    if (typeof res.flushHeaders !== 'function') {
      res.flushHeaders = function flushHeaders() {
        return res;
      };
    }
  }

  /**
   * Matches a route definition against a path.
   * @param {string[]} segments - Route segments.
   * @param {string} pathname - Request pathname.
   * @returns {object|null} Route parameters or null.
   */
  function matchRoute(segments, pathname) {
    const pathSegments = pathname.split('/').filter(Boolean);
    if (segments.length !== pathSegments.length) {
      return null;
    }

    const params = {};
    for (let index = 0; index < segments.length; index += 1) {
      const routeSegment = segments[index];
      const pathSegment = pathSegments[index];
      if (routeSegment.startsWith(':')) {
        params[routeSegment.slice(1)] = decodeURIComponent(pathSegment);
        continue;
      }

      if (routeSegment !== pathSegment) {
        return null;
      }
    }

    return params;
  }

  /**
   * Returns a MIME type for a file path.
   * @param {string} filePath - File path.
   * @returns {string} MIME type.
   */
  function contentTypeForFile(filePath) {
    const ext = path.extname(filePath).toLowerCase();
    if (ext === '.html') {
      return 'text/html; charset=utf-8';
    }
    if (ext === '.js' || ext === '.mjs') {
      return 'text/javascript; charset=utf-8';
    }
    if (ext === '.css') {
      return 'text/css; charset=utf-8';
    }
    if (ext === '.json') {
      return 'application/json; charset=utf-8';
    }
    return 'application/octet-stream';
  }

  /**
   * Express-style app function.
   * @param {http.IncomingMessage} req - Request.
   * @param {http.ServerResponse} res - Response.
   * @returns {Promise<void>}
   */
  async function app(req, res) {
    return handleRequest(req, res);
  }

  app.use = use;
  app.get = (routePath, handler) => registerRoute('GET', routePath, handler);
  app.post = (routePath, handler) => registerRoute('POST', routePath, handler);
  app.disable = disable;
  app.listen = listen;
  app.disabledSettings = disabledSettings;
  app.handle = handleRequest;

  return app;
}

/**
 * Creates JSON parsing middleware.
 * @returns {Function} Middleware.
 */
createExpress.json = function json() {
  return async function jsonMiddleware(req, res, next) {
    if (!shouldParseBody(req, 'application/json')) {
      req.body = req.body ?? {};
      return next();
    }

    req.body = await readRequestBody(req, parseJsonBody);
    return next();
  };
};

/**
 * Creates URL-encoded parsing middleware.
 * @returns {Function} Middleware.
 */
createExpress.urlencoded = function urlencoded() {
  return async function urlencodedMiddleware(req, res, next) {
    if (!shouldParseBody(req, 'application/x-www-form-urlencoded')) {
      req.body = req.body ?? {};
      return next();
    }

    req.body = await readRequestBody(req, parseUrlEncodedBody);
    return next();
  };
};

/**
 * Creates static file middleware.
 * @param {string} rootDir - Root directory.
 * @param {{ extensions?: string[] }} [options] - Static options.
 * @returns {Function} Middleware.
 */
createExpress.static = function staticMiddleware(rootDir, options = {}) {
  const extensions = Array.isArray(options.extensions) ? options.extensions : [];
  return function serveStatic(req, res, next) {
    if (req.method !== 'GET' && req.method !== 'HEAD') {
      return next();
    }

    const urlPath = decodeURIComponent((req.path || '/').replace(/^\/+/, ''));
    const candidates = [];

    if (!urlPath) {
      candidates.push(path.join(rootDir, 'index.html'));
    } else {
      candidates.push(path.join(rootDir, urlPath));
      for (const extension of extensions) {
        candidates.push(path.join(rootDir, `${urlPath}.${extension.replace(/^\./, '')}`));
      }
    }

    for (const candidate of candidates) {
      if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
        res.setHeader('Content-Type', candidate.endsWith('.html') ? 'text/html; charset=utf-8' : 'application/octet-stream');
        if (req.method === 'HEAD') {
          res.end();
        } else {
          res.end(fs.readFileSync(candidate));
        }
        return;
      }
    }

    return next();
  };
};

/**
 * Tests whether the request should be parsed.
 * @param {http.IncomingMessage} req - Request.
 * @param {string} contentTypeFragment - Content type fragment.
 * @returns {boolean} True when body parsing should run.
 */
function shouldParseBody(req, contentTypeFragment) {
  const contentType = String(req.headers['content-type'] || '').toLowerCase();
  return contentType.includes(contentTypeFragment);
}

/**
 * Reads a request body and parses it.
 * @param {http.IncomingMessage} req - Request.
 * @param {Function} parser - Body parser.
 * @returns {Promise<any>} Parsed body.
 */
function readRequestBody(req, parser) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on('data', (chunk) => chunks.push(Buffer.from(chunk)));
    req.on('end', () => {
      try {
        resolve(parser(Buffer.concat(chunks).toString('utf8')));
      } catch (error) {
        reject(error);
      }
    });
    req.on('error', reject);
  });
}

/**
 * Parses JSON request text.
 * @param {string} text - Raw text.
 * @returns {object} Parsed JSON body.
 */
function parseJsonBody(text) {
  if (!text) {
    return {};
  }
  return JSON.parse(text);
}

/**
 * Parses URL-encoded request text.
 * @param {string} text - Raw text.
 * @returns {object} Parsed form body.
 */
function parseUrlEncodedBody(text) {
  const result = {};
  for (const part of text.split('&')) {
    if (!part) {
      continue;
    }
    const [rawKey, rawValue = ''] = part.split('=');
    result[decodeURIComponent(rawKey.replace(/\+/g, ' '))] = decodeURIComponent(rawValue.replace(/\+/g, ' '));
  }
  return result;
}

module.exports = createExpress;
